---
title: "PRS Tutorial [R parts]"
output: html_document
date: "2023-11-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


    SOOOO, Reading in the height.gwas
    
> library(data.table)
> dat <- fread("Height.gwas.txt.gz") 
# copy it into R directory first
> result <- dat[INFO > 0.8 & MAF > 0.01]
# filters SNPs for MAF and imputation score [INFO]
> fwrite(result, "Height.gz", sep="\t")


    Target data QC after PLINK
    
> dat <- read.table("EUR.QC.het", header=T) # Read in the EUR.het file, specify it has header
m > <- mean(dat$F) # Calculate the mean  
> m
[1] 0.003034763
> s <- sd(dat$F) # Calculate the SD
> s
[1] 0.0138189
> valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
> write.table(valid[,c(1,2)], "EUR.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples


    Mismatched SNPs [usually done automatically by PRS software]

> bim <- read.table("EUR.bim")
> colnames(bim) <- c("CHR", "SNP", "CM", "BP", "B.A1", "B.A2")
# Read in QCed SNPs
> qc <- read.table("EUR.QC.snplist", header = F, stringsAsFactors = F)
> height <-
    read.table(gzfile("Height.QC.gz"),
            header = T,
            stringsAsFactors = F, 
            sep="\t")
> height$A1 <- toupper(height$A1)
height$A2 <- toupper(height$A2)
bim$B.A1 <- toupper(bim$B.A1)
bim$B.A2 <- toupper(bim$B.A2)
#changed to upper letters
# Merge summary statistic with target
> info <- merge(bim, height, by = c("SNP", "CHR", "BP"))
# Filter QCed SNPs
> info <- info[info$SNP %in% qc$V1,]
# Function for finding the complementary allele
> complement <- function(x) {
    switch (
        x,
        "A" = "T",
        "C" = "G",
        "T" = "A",
        "G" = "C",
        return(NA)
    )
}
# Get SNPs that have the same alleles across base and target
> info.match <- subset(info, A1 == B.A1 & A2 == B.A2)
# Identify SNPs that are complementary between base and target
> info$C.A1 <- sapply(info$B.A1, complement)
> info$C.A2 <- sapply(info$B.A2, complement)
> info.complement <- subset(info, A1 == B.A1 & A2 == B.A2)

# Update the complementary alleles in the bim file
# This allow us to match the allele in subsequent analysis
complement.snps <- bim$SNP %in% info.complement$SNP
bim[complement.snps,]$B.A1 <-
    sapply(bim[complement.snps,]$B.A1, complement)
bim[complement.snps,]$B.A2 <-
    sapply(bim[complement.snps,]$B.A2, complement)
# identify SNPs that need recoding
info.recode <- subset(info, A1 == B.A2 & A2 == B.A1)
# Update the recode SNPs
recode.snps <- bim$SNP %in% info.recode$SNP
tmp <- bim[recode.snps,]$B.A1
bim[recode.snps,]$B.A1 <- bim[recode.snps,]$B.A2
bim[recode.snps,]$B.A2 <- tmp
# identify SNPs that need recoding & complement
info.crecode <- subset(info, A1 == B.A2 & A2 == B.A1)
# Update the recode + strand flip SNPs
com.snps <- bim$SNP %in% info.crecode$SNP
tmp <- bim[com.snps,]$B.A1
bim[com.snps,]$B.A1 <- as.character(sapply(bim[com.snps,]$B.A2, complement))
bim[com.snps,]$B.A2 <- as.character(sapply(tmp, complement))
# Output updated bim file
write.table(
    bim[,c("SNP", "B.A1")],
    "EUR.a1",
    quote = F,
    row.names = F,
    col.names = F,
    sep="\t"
)
# Identify SNPs that have different allele in base and target (usually due to difference in genome build or Indel)
mismatch <-
    bim$SNP[!(bim$SNP %in% info.match$SNP |
                bim$SNP %in% info.complement$SNP | 
                bim$SNP %in% info.recode$SNP |
                bim$SNP %in% info.crecode$SNP)]
write.table(
    mismatch,
    "EUR.mismatch",
    quote = F,
    row.names = F,
    col.names = F
)


     sex check needs terminal & plink, then R for last bits

./plink \
    --bfile EUR \
    --extract EUR.QC.prune.in \
    --keep EUR.valid.sample \
    --check-sex \
    --out EUR.QC
# make sure all files are in the right directory
# R to finish off with the F stats
library(data.table)
# Read in file
valid <- fread("EUR.valid.sample")
dat <- fread("EUR.QC.sexcheck")[FID%in%valid$FID]
fwrite(dat[STATUS=="OK",c("FID","IID")], "EUR.QC.valid", sep="\t") 
# made a table EUR.QC.valid for matched reported and bio sexes

